File: src/lib/accountInstance.ts
// ============================================================================
// DrKLO/Telegram TMessagesProj: AccountInstance.java Architecture
// Source: TMessagesProj/src/main/java/org/telegram/messenger/AccountInstance.java
// ============================================================================

import { UserConfig } from './userConfig';
import { MessagesController } from './messagesController';
import { MessagesStorage } from './messagesStorage';
import { ConnectionsManager } from './connectionsManager';

export class AccountInstance {
  private static instances: AccountInstance[] = new Array(UserConfig.MAX_ACCOUNT_COUNT);
  private currentAccount: number;

  constructor(instance: number) {
    this.currentAccount = instance;
  }

  public static getInstance(num: number = 0): AccountInstance {
    const idx = Math.max(0, Math.min(num, UserConfig.MAX_ACCOUNT_COUNT - 1));
    if (!AccountInstance.instances[idx]) {
      AccountInstance.instances[idx] = new AccountInstance(idx);
    }
    return AccountInstance.instances[idx];
  }

  public getCurrentAccount(): number {
    return this.currentAccount;
  }

  public getUserConfig(): UserConfig {
    return UserConfig.getInstance(this.currentAccount);
  }

  public getMessagesController(): MessagesController {
    return MessagesController.getInstance(this.currentAccount);
  }

  public getMessagesStorage(): MessagesStorage {
    return MessagesStorage.getInstance(this.currentAccount);
  }

  public getConnectionsManager(): ConnectionsManager {
    return ConnectionsManager.getInstance(this.currentAccount);
  }

  /**
   * Switches the active account, initializes its controller, and loads only its real dialogs
   */
  public static setAsPrimaryAccount(realAccountNum: number) {
    UserConfig.selectedAccount = realAccountNum;
    const realInstance = AccountInstance.getInstance(realAccountNum);
    realInstance.getMessagesController().loadDialogs(0, 100, true);
  }
}

export const accountInstance = AccountInstance.getInstance(0);

File: TMessagesProj/src/main/java/org/telegram/messenger/AccountInstance.java
/*
 * This is the source code of Telegram for Android v. 11.x.x.
 * It is licensed under GNU GPL v. 2 or later.
 * You should have received a copy of the license in this archive (see LICENSE).
 *
 * Copyright Nikolai Kudashov, 2013-2024.
 */

package org.telegram.messenger;

import org.telegram.tgnet.ConnectionsManager;

public class AccountInstance {
    public static final int MAX_ACCOUNT_COUNT = 4;
    private static volatile AccountInstance[] Instance = new AccountInstance[MAX_ACCOUNT_COUNT];

    private final int currentAccount;

    public static AccountInstance getInstance(int num) {
        AccountInstance localInstance = Instance[num];
        if (localInstance == null) {
            synchronized (AccountInstance.class) {
                localInstance = Instance[num];
                if (localInstance == null) {
                    Instance[num] = localInstance = new AccountInstance(num);
                }
            }
        }
        return localInstance;
    }

    public AccountInstance(int instance) {
        currentAccount = instance;
    }

    public int getCurrentAccount() {
        return currentAccount;
    }

    public MessagesController getMessagesController() {
        return MessagesController.getInstance(currentAccount);
    }

    public MessagesStorage getMessagesStorage() {
        return MessagesStorage.getInstance(currentAccount);
    }

    public ConnectionsManager getConnectionsManager() {
        return ConnectionsManager.getInstance(currentAccount);
    }

    public NotificationCenter getNotificationCenter() {
        return NotificationCenter.getInstance(currentAccount);
    }

    public UserConfig getUserConfig() {
        return UserConfig.getInstance(currentAccount);
    }
}
